server/src/domain/media/media.util.ts

```diff
@@ -0,0 +1,199 @@
+      // the file for improved playback speed.
+      '-movflags faststart',
+      '-fps_mode passthrough',
+    ];
+  }
+
+  getFilterOptions(stream: VideoStreamInfo) {
+    const options = [];
+    if (shouldScale(stream, this.config)) {
+      options.push(`scale=${getScaling(stream, this.config)}`);
+    }
+
+    return options;
+  }
+
+  getPresetOptions() {
+    return [`-preset ${this.config.preset}`];
+  }
+
+  abstract getBitrateOptions(): Array<string>;
+  abstract getThreadOptions(): Array<string>;
+}
+
+export class H264Handler extends BaseHandler implements VideoCodecSWHandler {
+  getBitrateOptions() {
+    const bitrates = getBitrateDistribution(this.config);
+    if (eligibleForTwoPass(this.config)) {
+      return [
+        `-b:v ${bitrates.target}${bitrates.unit}`,
+        `-minrate ${bitrates.min}${bitrates.unit}`,
+        `-maxrate ${bitrates.max}${bitrates.unit}`,
+      ];
+    } else if (bitrates.max > 0) {
+      // -bufsize is the peak possible bitrate at any moment, while -maxrate is the max rolling average bitrate
+      // needed for -maxrate to be enforced
+      return [
+        `-crf ${this.config.crf}`,
+        `-maxrate ${bitrates.max}${bitrates.unit}`,
+        `-bufsize ${bitrates.max * 2}${bitrates.unit}`,
+      ];
+    } else {
+      return [`-crf ${this.config.crf}`];
+    }
+  }
+
+  getThreadOptions() {
+    if (this.config.threads <= 0) {
+      return [];
+    }
+    return [
+      `-threads ${this.config.threads}`,
+      '-x264-params "pools=none"',
+      `-x264-params "frame-threads=${this.config.threads}"`,
+    ];
+  }
+}
+
+export class HEVCHandler extends H264Handler {
+  getThreadOptions() {
+    if (this.config.threads <= 0) {
+      return [];
+    }
+    return [
+      `-threads ${this.config.threads}`,
+      '-x265-params "pools=none"',
+      `-x265-params "frame-threads=${this.config.threads}"`,
+    ];
+  }
+}
+
+export class VP9Handler extends BaseHandler implements VideoCodecSWHandler {
+  getPresetOptions() {
+    const speed = Math.min(getPresetIndex(this.config.preset), 5); // values over 5 require realtime mode, which is its own can of worms since it overrides -crf and -threads
+    if (speed >= 0) {
+      return [`-cpu-used ${speed}`];
+    }
+    return [];
+  }
+
+  getBitrateOptions() {
+    const bitrates = getBitrateDistribution(this.config);
+    if (eligibleForTwoPass(this.config)) {
+      return [
+        `-b:v ${bitrates.target}${bitrates.unit}`,
+        `-minrate ${bitrates.min}${bitrates.unit}`,
+        `-maxrate ${bitrates.max}${bitrates.unit}`,
+      ];
+    }
+
+    return [`-crf ${this.config.crf}`, `-b:v ${bitrates.max}${bitrates.unit}`];
+  }
+
+  getThreadOptions() {
+    const options = ['-row-mt 1'];
+    if (this.config.threads) {
+      options.push(`-threads ${this.config.threads}`);
+    }
+    return options;
+  }
+}
+
+export function eligibleForTwoPass(config: SystemConfigFFmpegDto) {
```

Comment:
All these methods should be on the base class

Reply:
Good thinking. These are from before I made the class handlers so they stayed around as regular functions.